go/hyperkit.go

```
@@ -0,0 +1,524 @@
+		a = append(a, "-s", fmt.Sprintf("4,ahci-cd,%s", h.ISOImage))
+	}
+
+	a = append(a, "-s", "10,virtio-rnd")
+	a = append(a, "-s", "31,lpc")
+
+	if h.Console == ConsoleFile {
+		a = append(a, "-l", fmt.Sprintf("com1,autopty=%s/tty,log=%s/console-ring", h.StateDir, h.StateDir))
+	} else {
+		a = append(a, "-l", "com1,stdio")
+	}
+
+	kernArgs := fmt.Sprintf("kexec,%s,%s,earlyprintk=serial %s", h.Kernel, h.Initrd, cmdline)
+	a = append(a, "-f", kernArgs)
+
+	h.Arguments = a
+	h.CmdLine = h.HyperKit + " " + strings.Join(a, " ")
+}
+
+// Execute hyperkit and plumb stdin/stdout/stderr.
+func (h *HyperKit) execHyperKit() error {
+
+	cmd := exec.Command(h.HyperKit, h.Arguments...)
+	cmd.Env = os.Environ()
+
+	// Plumb in stdin/stdout/stderr. If ConsoleStdio is configured
+	// plumb them to the system streams. If a logger is specified,
+	// use it for stdout/stderr logging. Otherwise use the default
+	// /dev/null.
+	if h.Console == ConsoleStdio {
+		cmd.Stdin = os.Stdin
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stderr
+	} else if h.log != nil {
+		stdoutChan := make(chan string)
+		stderrChan := make(chan string)
+		stdout, err := cmd.StdoutPipe()
+		if err != nil {
+			return err
+		}
+		stderr, err := cmd.StderrPipe()
+		if err != nil {
+			return err
+		}
+		stream(stdout, stdoutChan)
+		stream(stderr, stderrChan)
+
+		done := make(chan struct{})
+		go func() {
+			for {
+				select {
+				case stderrl := <-stderrChan:
+					log.Printf("Hyperkit ERR: ", stderrl)
+				case stdoutl := <-stdoutChan:
+					log.Printf("Hyperkit OUT: ", stdoutl)
+				case <-done:
+					return
+				}
+			}
+		}()
+	}
+
+	err := cmd.Start()
+	if err != nil {
+		return err
+	}
+	h.Pid = cmd.Process.Pid
+	h.process = cmd.Process
+	h.writeState()
+	if !h.background {
+		err = cmd.Wait()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// writeState write the state to a JSON file
+func (h *HyperKit) writeState() {
+	if h.StateDir == "" {
+		return
+	}
+
+	s, _ := json.Marshal(h)
```

You should probably propagate this error, or else other things might get quite confused trying to unpickle it later.